src/Agent.Worker/StepsRunner.cs

```diff
@@ -205,5 +204,36 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 Trace.Info($"Current state: job state = '{jobContext.Result}', step failed = {stepFailed}, critical step failed = {criticalStepFailed}");
             }
         }
+
+        private bool ShouldRunStep(IStep step, bool stepFailed, bool criticalStepFailed)
+        {            
+            switch(step.RunMode)
+            {
+                // The server hasn't been upgraded yet. Fallback to old code
+                case RunMode.Undefined:
```

Comment:
It can be done by making the inference when data is being populated into TaskRunner from TaskInstance
`    public sealed class TaskRunner : AgentService, ITaskRunner
    {
        public bool AlwaysRun => TaskInstance?.AlwaysRun ?? default(bool);
        public RunMode RunMode => TaskInstance?.RunMode ?? RunMode.Default;
        public IList<TaskCondition> Conditions => TaskInstance?.Conditions ?? new List<TaskCondition>();
        public bool ContinueOnError => TaskInstance?.ContinueOnError ?? default(bool);
        public bool Critical => false;
        public string DisplayName => TaskInstance?.DisplayName;
        public bool Enabled => TaskInstance?.Enabled ?? default(bool);
        public IExecutionContext ExecutionContext { get; set; }
        public bool Finally => false;
        public TaskInstance TaskInstance { get; set; }
    }`